cmd/collector/app/collector.go

```diff
@@ -0,0 +1,183 @@
+	ServiceName    string
+	Logger         *zap.Logger
+	MetricsFactory metrics.Factory
+	SpanWriter     spanstore.Writer
+	StrategyStore  strategystore.StrategyStore
+	HealthCheck    *healthcheck.HealthCheck
+}
+
+// New constructs a new collector component, ready to be started
+func New(params *CollectorParams) *Collector {
+	return &Collector{
+		serviceName:    params.ServiceName,
+		logger:         params.Logger,
+		metricsFactory: params.MetricsFactory,
+		spanWriter:     params.SpanWriter,
+		strategyStore:  params.StrategyStore,
+		hCheck:         params.HealthCheck,
+	}
+}
+
+// Start the component and underlying dependencies
+func (c *Collector) Start(builderOpts *CollectorOptions) error {
+	handlerBuilder := &SpanHandlerBuilder{
+		SpanWriter:     c.spanWriter,
+		CollectorOpts:  *builderOpts,
+		Logger:         c.logger,
+		MetricsFactory: c.metricsFactory,
+	}
+	zipkinSpansHandler, jaegerBatchesHandler, grpcHandler := handlerBuilder.BuildHandlers()
+	recoveryHandler := recoveryhandler.NewRecoveryHandler(c.logger, true)
+
+	if tchServer, err := server.StartThriftServer(&server.ThriftServerParams{
+		ServiceName:          c.serviceName,
+		Port:                 builderOpts.CollectorPort,
+		JaegerBatchesHandler: jaegerBatchesHandler,
+		ZipkinSpansHandler:   zipkinSpansHandler,
+		StrategyStore:        c.strategyStore,
+		Logger:               c.logger,
+	}); err != nil {
+		c.logger.Fatal("Could not start Thrift collector", zap.Error(err))
+	} else {
+		c.tchServer = tchServer
+	}
+
+	if grpcServer, err := server.StartGRPCServer(&server.GRPCServerParams{
+		Port:          builderOpts.CollectorGRPCPort,
+		Handler:       grpcHandler,
+		TLSConfig:     builderOpts.TLS,
+		SamplingStore: c.strategyStore,
+		Logger:        c.logger,
+	}); err != nil {
+		c.logger.Fatal("Could not start gRPC collector", zap.Error(err))
+	} else {
+		c.grpcServer = grpcServer
+	}
+
+	if httpServer, err := server.StartHTTPServer(&server.HTTPServerParams{
+		Port:            builderOpts.CollectorHTTPPort,
+		Handler:         jaegerBatchesHandler,
+		RecoveryHandler: recoveryHandler,
+		HealthCheck:     c.hCheck,
+		MetricsFactory:  c.metricsFactory,
+		SamplingStore:   c.strategyStore,
+		Logger:          c.logger,
+	}); err != nil {
+		c.logger.Fatal("Could not start the HTTP server", zap.Error(err))
+	} else {
+		c.hServer = httpServer
+	}
+
+	if zkServer, err := server.StartZipkinServer(&server.ZipkinServerParams{
+		Port:            builderOpts.CollectorZipkinHTTPPort,
+		Handler:         zipkinSpansHandler,
+		RecoveryHandler: recoveryHandler,
+		AllowedHeaders:  builderOpts.CollectorZipkinAllowedHeaders,
+		AllowedOrigins:  builderOpts.CollectorZipkinAllowedOrigins,
+		Logger:          c.logger,
+	}); err != nil {
+		c.logger.Fatal("Could not start the Zipkin server", zap.Error(err))
+	} else {
+		c.zkServer = zkServer
+	}
+
+	return nil
+}
+
+// Close the component and all its underlying dependencies
+func (c *Collector) Close() error {
```

Comment:
The return type is part of the [`io.Closer`](https://golang.org/pkg/io/#Closer) interface.

Reply:
Then return the error. There are several return assignments in this function and none error being returned.

Reply:
There's no error that we would want to bubble up from here. For instance, when issuing a `Close()`, we don't want to break the routine when the HTTP server failed to close, as that would result in the Zipkin server not being closed.

Reply:
It does not mean that the error has to be returned _immediatelly_. It can be wrapped and returned at the end of the function. Hopefully, this function is not used by many clients...

Reply:
Interesting idea. If the HTTP server and Zipkin fails, Zipkin will end up wrapping the HTTP server failure. Not sure that's intuitive from the perspective of the admin looking at the logs.

> Hopefully, this function is not used by many clients...

It's being used by the `main.go` of both the Collector and All In One.